### PR TITLE
Change the `_compute_flux_spatial` method on `MapEvaluator`

### PR DESCRIPTION
**Pull request by fabiopintore**
This PR includes some changes in the `_compute_flux_spatial` function in `~gammapy.datasets.map.MapEvaluator`. 
We now allows the function to calculate the integral flux over either a `WcsGeom` or a `RegionNDMap`. The integrated flux is also corrected for the PSF.

To do:
1) We need to add tests!
2) possibility to enlarge the tangential geometry so that the PSF is fully contained.

@LauraOlivera

### CODE REVIEW

File: gammapy/datasets/map.py [hunk 1]
```diff
@@ -2571,17 +2575,37 @@ def compute_flux_psf_convolved(self):
 
         if self.model.spatial_model and not isinstance(self.geom, RegionGeom):
             value = value * self.compute_flux_spatial().quantity
+        else:
```
Comment (adonath) [inline gammapy/datasets/map.py:2578]: Just remove the `not isinstance(self.geom, RegionGeom)` before...

File: gammapy/datasets/map.py [hunk 2]
```diff
@@ -2571,17 +2575,37 @@ def compute_flux_psf_convolved(self):
...
+
+        Returns
+        ----------
+        value: `~astropy.units.Quantity`
+            Psf-corrected, integrated flux over a given region.
+        """
+        if isinstance(self.geom, RegionGeom):
+            wcs_geom = self.geom.to_wcs_geom()
+            values = self.model.spatial_model.integrate_geom(wcs_geom)
+            tmp_map = Map.from_geom(geom=wcs_geom, data=values.data,
```
Comment (adonath) [inline gammapy/datasets/map.py:2597]: `.interate_geom()` already returns a `Map` object...no need to create it here...

File: gammapy/datasets/map.py [hunk 3]
```diff
@@ -2571,17 +2575,37 @@ def compute_flux_psf_convolved(self):
...
+                                    unit=values.unit)
+
+            if self.psf and self.model.apply_irf["psf"]:
+                values = self.apply_psf(tmp_map)
+            value = ((values.data * wcs_geom.solid_angle())).sum()
```
Comment (adonath) [inline gammapy/datasets/map.py:2602]: This misses the application of the region mask, that you had in your previous implementation. As you have called `.integrate_geom()` already, there is no need to multiply again with the solid angle...

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @fabiopintore, I have left a few comments. Please add a unit test now to check the implementation...

File: gammapy/datasets/tests/test_map.py [hunk 4]
```diff
@@ -1361,3 +1361,46 @@ def test_downsample_onoff():
...
+    background_region = RegionNDMap.create(region, axes=[energy_axis_true])
+    background_region.data += 2
+
+    dataset_region = MapDataset(psf=psf, exposure=exposure_region, background=background_region)
+    dataset_region.models = [model]
+
+    npred_region = dataset_region.npred()
+
+    # for a WCSmap
+    exposure_wcs = Map.create(axes=[energy_axis_true], region=region, map_type="region")
```
Comment (adonath) [inline gammapy/datasets/tests/test_map.py:1395]: Note this does not create a `WCSMap`, but again a `RegionMap`, for a WCS map you need to choose `map_type="wcs"`...please adapt the test.

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @fabiopintore, I would suggest to add a second test using a circular region of the size  of 1 sigma of the Gaussian PSF and check the spread corresponds to ~39%.

**Comment by fabiopintore** [comment]
I added a further test to check if a circular region of the 1sigma value of the PSF contains the correct amount of flux. 
@adonath, @LauraOlivera  please may you check if I made the right things? The tests pass, but it seems to me that the predicted counts, when using a `RegionGeom`, are too much than expected for the given exposure...

File: gammapy/datasets/map.py [hunk 5]
```diff
@@ -2569,19 +2573,36 @@ def compute_flux_psf_convolved(self):
...
+            Psf-corrected, integrated flux over a given region.
+        """
+        if isinstance(self.geom, RegionGeom):
+            wcs_geom = self.geom.to_wcs_geom()
+            mask = self.geom.contains(wcs_geom.get_coord())
+            values = self.model.spatial_model.integrate_geom(wcs_geom)
+
+            if self.psf and self.model.apply_irf["psf"]:
+                values = self.apply_psf(values)
+            value = (values.data[mask]).sum()
```
Comment (adonath) [inline gammapy/datasets/map.py:2599]: I started to test implementation and I noticed a bug in this line: we have to sum only over the spatial axes here and not all axes (and not including the energy axis). So the statement here should be something like `.sum(axis=(1, 2))`... I need to check whether the implementation in `SpatialModel.integrate_geom()` is correct.

File: gammapy/datasets/tests/test_map.py [hunk 6]
```diff
@@ -1361,3 +1361,67 @@ def test_downsample_onoff():
     assert downsampled.counts.data.sum() == dataset_onoff.counts.data.sum()
     assert downsampled.counts_off.data.sum() == dataset_onoff.counts_off.data.sum()
     assert_allclose(downsampled.alpha.data, 0.5)
+
+
+def test_compute_flux_spatial():
```
Comment (adonath) [inline gammapy/datasets/tests/test_map.py:1366]: I realised this test case is maybe a bit to high level. What about if we start on the `MapEvaluator` level?
So define:
- Constant exposure (filled with ones...) with 2 energy bins
- Point source models with a `ConstSpectralModel`
- Gaussian PSF kernel, created with `PSFKernel.from_gauss(geom=region_geom.to_wc_geom())`
- Create a `MapEvaluator()`
- `MapEvaluator.compute_flux_spatial()` should then return exactly the PSF containment per energy as an array like [0.39, 0.39]...

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @fabiopintore, I noticed a bug in the implementation and made a proposal for a better unit test...

**Review by adonath — APPROVED** [review]
Thanks @fabiopintore!

**Pull request merged**